helix-tui/src/text.rs

```diff
@@ -229,6 +229,70 @@ impl<'a> Spans<'a> {
     pub fn width(&self) -> usize {
         self.0.iter().map(Span::width).sum()
     }
+
+    /// Constructs a styled `Span` from the given `string` and `style`
+    /// and appends it the to the `Spans`.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// # use helix_tui::text::{Span, Spans};
+    /// # use helix_view::graphics::{Color, Style};
+    /// let mut spans = Spans::new();
+    /// spans.push("Hello!", Style::default());
+    /// assert_eq!("Hello!", spans.0[0].content);
+    /// ```
+    pub fn push<T>(&mut self, string: T, style: Style)
+    where
+        T: Into<Cow<'a, str>>,
+    {
+        self.0.push(Span::styled(string, style));
+    }
+
+    /// Creates a new, empty `Spans`.
+    pub fn new() -> Spans<'a> {
+        Spans(vec![])
+    }
+
+    /// Constructs a single, styled `Span` with the given `string` and `style`
+    /// and wraps it in a `Spans`.
+    pub fn single<T>(string: T, style: Style) -> Spans<'a>
+    where
+        T: Into<Cow<'a, str>>,
+    {
+        Spans(vec![Span::styled(string, style)])
+    }
+
+    /// Intersperses `sep` between items of the given iterator then returns the concatenated Spans.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// # use helix_tui::text::{Span, Spans};
+    /// # use helix_view::graphics::Style;
+    /// let spans = Spans::join(
+    ///   vec![
+    ///     Spans::single("Hello", Style::default()),
+    ///     Spans::single("World!", Style::default()),
+    ///   ].into_iter(),
+    ///   &Spans::single(" ", Style::default()),
+    /// );
+    /// assert_eq!("Hello World!", String::from(spans));
+    /// ```
+    pub fn join<S>(mut spanses: S, sep: &Spans<'a>) -> Spans<'a>
```

Comment:
```suggestion
    pub fn join<S>(mut spans_iter: S, sep: &Spans<'a>) -> Spans<'a>
```

Reply:
Does `sep` need to be a `Spans` here? I think it could just be a `Span`.

Reply:
Using spans allows for fancier separators, for example one with multiple colors.

I think I will make two more versions of `join`, one taking a `Span` as you suggest, and one taking an interpolator function which takes two neighboring `Spans` and generates a separator for them. This could be useful for powerline-style color transitions.

Reply:
Variable naming fixed in 8e8d860b

Reply:
Removed all custom Spans functions so this issue is now resolved.